### DISCUSSION TITLE
setting safelist in tailwind.config.js consumes a lot of CPU usage and memory

### DISCUSSION THREAD

**Discussion by silenius** [tailwindlabs/tailwindcss#13376] [Help]
Hello,

I'm currently developing a WYSIWYG editor based on TipTap with TailwindCSS integration (with VueJS 3, Vite). The user could potentially set any TailwindCSS class so I've to include several classes by default in the build. 
I've already built the UI for the _padding_, the _text-color_ and the _background color_, so I've added some patterns in `safelist`:

Currently I have something like:
```javascript
~/code/projects/amnesia_admin/ cat tailwind.config.js
module.exports = {
  content: [
    "./index.html",
    "./src/**/*.{vue,js,ts,jsx,tsx}",
  ],
  theme: {
    extend: {},
  },
  plugins: [
    require('@tailwindcss/forms'),
    require('@tailwindcss/typography'),
  ],
  safelist: [
    {
      pattern: /(h|w|px|py|pt|pb|pl|pr)-\d+/,
      variants: ['hover', 'sm', 'md', 'lg', 'xl', '2xl']
    },
    {
      pattern: /(text|bg|outline)-/,
      variants: ['hover', 'sm', 'md', 'lg', 'xl', '2xl']
    }
  ]
}
```
However, things are slow and unusable, node consumes 100%CPU for a minute or so and takes 1.5GB of RAM (dev server).

How could I solve this? I was wondering if there is a better approach?

Thanks!

**Comment by RobinMalfait** [comment]
Hey!

The problem here are the safelist patterns you are using, those regexes match _a lot_ of utilities (about 157 thousand). These utilities alone generate 13 MB worth of CSS... you don't want that.

Allowing _any_ Tailwind class is unfortunately not going to work as expected because then you need to generate even more CSS which will take an even longer time and your browser is not going to be happy about that unfortunately.

I don't have a perfect solution for this scenario, but a few things I can think of that might give you some pointers on how to handle this situation are:

1. Don't accept _all_ Tailwind classes but just a subset — this way you can control the exact Tailwind classes that are generated.
2. If the WYSIWYG editor for let's say an `h1` generates a style for `color: red` you _could_ map these to CSS variables:
   ```html
   <div style="--my-text-color: {{ color }}" class="text-[var(--my-text-color)]"></div>
   ```
   This way you only have to generate CSS for the `text-[var(--my-text-color)]` class and the value itself can be fully dynamic.
3. An even crazier idea would be to run Tailwind CSS on the server side given the contents of the WYSIWYG editor, then store the generated CSS in a database next to the actual contents. This would only work if the WYSIWYG editor is coupled to a CMS of some sorts. If you are developing a package that exposes the WYSIWYG editor as a component, then this would not be possible unfortunately.

So long story short, I don't have the perfect solution for this scenario, but the main goal should be to reduce the amount of generated CSS because even if you generate _all_ available classes the browser won't be happy about that.

Hope this helps!

**Reply by silenius** [reply]
Thanks, option 3 might be feasible as the editor is indeed only a part of our internal CMS. What do you mean exactly by "running TailwindCSS on the server side" ?
Another option would be to use the CDN..?
